rust-windowing/winit
# Impl. page CSS layout-based canvas sizing (web-sys)

- [ ] Compilation warnings were addressed
- [ ] `cargo fmt` has been run on this branch
- [ ] `cargo doc` builds successfully
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Here is a POC implementation for #1661.

I referred to the code in https://github.com/KhronosGroup/WebGL/issues/2460#issuecomment-625045885 and implemented it with `web-sys`. It basically supports three modes (listed in order of preference):

- Use `ResizeObserver` with `device-pixel-content-box` to get the real device pixel dimensions (supported in Chrome 84 or above only)
- Use `ResizeObserver` to detect change in size with window resize event for backup, accompanied with some code to attempt to snap the canvas to device pixel (works on most recent browsers)
- Use window resize event only to check for change in size (would fail if the page layout changed without a window resize)

I have tested extensively on Chrome 85 beta and Firefox 80 beta on Windows and was able to get a crisp rendering (using a render of 1px lines, I can share the code if anyone wants to see but it's mostly similar to the test code on the linked WebGL issue thread) on different zoom levels and high-dpi display scaling. The canvas reacts to layout changes nicely (I tested with flexbox). I also tested that the fallback without `ResizeObserver` is functional.

### Technical details

- Since `web-sys` does not yet have the binding to `ResizeObserver`, I followed [these instructions](https://rustwasm.github.io/wasm-bindgen/contributing/web-sys/supporting-more-web-apis.html?highlight=webidl#supporting-more-web-apis-in-web-sys) to generate the bindings from the webidl and included them in a commit.
- Since there are multiple events that could cause the canvas to be resized and to also keep track of the scaling change, I made a `CanvasResizeObserver` struct to handle and emit the events corresponding to canvas size and scaling changes. this struct is currently quite a mess due to the need to handle multiple fallbacks and event listeners. I hope that a bit of refactoring will make it nicer.
- I am currently assuming that a change in scale factor will always come together with a size change.
- This PR builds upon PR #1659, but this PR doesn't really depend on its changes.
- To make this work, the canvas must be wrapped in a `<div>` element which must have a non-default CSS `position` (`position: relative` works). This parent element can be laid out normally.
- I added a `with_auto_parent_size` builder method. When used, the canvas will not be resizeable with user code.
